src/robusta/core/triggers/pod_crash_loop_trigger.py

```diff
@@ -0,0 +1,74 @@
+from ..discovery.top_service_resolver import TopServiceResolver
+from ...core.playbooks.base_trigger import TriggerEvent
+from ...integrations.kubernetes.autogenerated.triggers import PodUpdateTrigger, PodChangeEvent
+from ...integrations.kubernetes.base_triggers import K8sTriggerEvent
+from ...utils.rate_limiter import RateLimiter
+
+
+class PodCrashLoopTrigger(PodUpdateTrigger):
+    rate_limit: int = 3600
+    restart_reason: str = None
+    restart_count: int = 2
+
+    """
+    :var rate_limit: Limit firing to once every `rate_limit` seconds 
+    :var restart_reason: Limit restart loops for this specific reason. If omitted, all restart reasons will be included.
+    :var restart_count: Fire only after the specified number of restarts 
+    """
+
+    def __init__(
+            self,
+            name_prefix: str = None,
+            namespace_prefix: str = None,
+            labels_selector: str = None,
+            rate_limit: int = 3600,
+            restart_reason: str = None,
+            restart_count: int = 2,
+    ):
+        super().__init__(
+            name_prefix=name_prefix,
+            namespace_prefix=namespace_prefix,
+            labels_selector=labels_selector,
+        )
+        self.rate_limit = rate_limit
+        self.restart_reason = restart_reason
+        self.restart_count = restart_count
+
+    def should_fire(self, event: TriggerEvent, playbook_id: str):
+        should_fire = super().should_fire(event, playbook_id)
+        if not should_fire:
+            return should_fire
+
+        if not isinstance(event, K8sTriggerEvent):
+            return False
+
+        exec_event = self.build_execution_event(event, {})
+
+        if not isinstance(exec_event, PodChangeEvent):
+            return False
+
+        pod = exec_event.get_pod()
+
+        all_statuses = pod.status.containerStatuses + pod.status.initContainerStatuses
+        crashing = [
+            container_status
+            for container_status in all_statuses
+            if container_status.state.waiting is not None
+            and container_status.restartCount >= self.restart_count  # report only after the restart_count restart
+            and (self.restart_reason is None or self.restart_reason in container_status.state.waiting.reason)
+        ]
+
+        if not crashing:
+            return False
+
+        # Perform a rate limit for this service key according to the rate_limit parameter
+        name = pod.metadata.name
+        namespace = pod.metadata.namespace
+        service_key = TopServiceResolver.guess_service_key(
+            name=name, namespace=namespace
+        )
+        return RateLimiter.mark_and_test(
+            f"PodCrashLoopTrigger_{playbook_id}",
+            service_key if service_key else namespace + ":" + name,
```

Comment:
The logic here is different than before. I'm not sure we want to really rate-limit by the `service_key` because of the following scenario:

1. You update a deployment and something crashes.
2. You get a notification.
3. You fix the deployment and deploy again.
4. It crashes again before your fix wasn't good enough.
5. Now you are not notified.

IMO this is confusing and not the desired behaviour. What if we rate-limited by the replicaset name, which is available in `metadata.ownerReferences.name`? I think that would give the desitred behaviour as the replicaset name changes in each deployment update.

Reply:
you're right, fixed